### PR TITLE
Support for providing sense data for REQUEST SENSE

### PR DESCRIPTION
**Pull request by uweseimet**
These changes extend Error() with two optional arguments for the Sense Key and the Additional Sense Code and extend CmdRequestSense to return the values set by the previous (failed) command.
Unsupported SCSI commands now return the correct SCSI sense data, which improves compatibility with initiators evaluating these data.
In theory the sense data are not optional, but the current code is not prepared for providing these data for each error. Any command that fails should set these values according to the SCSI specs.

This PR resolves issues https://github.com/akuker/RASCSI/issues/116,  https://github.com/akuker/RASCSI/issues/119 and https://github.com/akuker/RASCSI/issues/124.

@akuker I successfully tested these changes with my SCSI hardware. I also tried to run the robot tests, but had an issue with the test suite reporting "Authentication failed for user 'pi'", so that the test suite could not be run. Any idea what might be wrong here?

### CODE REVIEW

**Comment by uweseimet** [comment]
Commit 0135abf resolves issue #124.

**Comment by akuker** [comment]
Hi @uweseimet - The Robot tests expect the RaSCSI to be accessible through the network and to have the default username/password. I'm guessing your RaSCSI device has a different password set.

File: src/raspberrypi/devices/disk.cpp [hunk 1]
```diff
@@ -2272,8 +2275,12 @@ int FASTCALL Disk::ReadCapacity(const DWORD* /*cdb*/, BYTE *buf)
 		return 0;
 	}
 
+	if (disk.blocks <= 0) {
+		// Capacity not available, medium not present
```
Comment (akuker) [inline src/raspberrypi/devices/disk.cpp:2279]: Log an error here?

File: src/raspberrypi/controllers/sasidev_ctrl.cpp [hunk 2]
```diff
@@ -1332,8 +1367,8 @@ void FASTCALL SASIDEV::CmdInvalid()
 	ASSERT(this);
 	LOGWARN("%s Command not supported", __PRETTY_FUNCTION__);
 
-	// Failure (Error)
-	Error();
+	// INVALID COMMAND
+	Error(0x05, 0x20);
```
Comment (akuker) [inline src/raspberrypi/controllers/sasidev_ctrl.cpp:1371]: Could you define these Sense Keys and ASC codes in a header file somewhere? I'd prefer to not have magic numbers without having some traceability to where they came from.

Comment (uweseimet) [reply src/raspberrypi/controllers/sasidev_ctrl.cpp]: Yes, that would be better, indeed. I added enums based on the official data from https://www.t10.org/lists/1spc-lst.htm to scsi.h.

File: src/raspberrypi/devices/disk.cpp (hunk 1 above)
Comment (uweseimet) [reply src/raspberrypi/devices/disk.cpp]: Done

**Comment by uweseimet** [comment]
@akuker Ready for merging, successfully tested with my hardware after resolving the conflicts.

**Comment by uweseimet** [comment]
> Hi @uweseimet - The Robot tests expect the RaSCSI to be accessible through the network and to have the default username/password. I'm guessing your RaSCSI device has a different password set.

After changing the password I could proceed. I had to install lsscsi (updated the Wiki accordingly), but now the tests fail because no SCSI devices (/dev/sd*) can be found.

--
==============================================================================
Robot                                                                         
==============================================================================
Robot.Rascsi Cd-Rom Functionality :: Test the RaSCSI CD-ROM emulation funct...
==============================================================================
MacOS formated ISO is mounted and correct size is reported :: Moun... bash: /sys/class/scsi_host/host0/scan: No such file or directory
--

After adding a USB stick /dev/sda and /dev/sdb are available, but the tests still fail:

--
==============================================================================
Robot                                                                         
==============================================================================
Robot.Rascsi Cd-Rom Functionality :: Test the RaSCSI CD-ROM emulation funct...
==============================================================================
MacOS formated ISO is mounted and correct size is reported :: Moun... | FAIL |
Unable to find device ID for SCSI ID 6: '' does not match '/dev/[\w\d]+'
--

Can it be that the tests expect a very specific device setup?

**Comment by akuker** [comment]
You'll need to download some specific files on your RaSCSI host. 

In the /home/pi/images folder, you'll need the following:
- Sim Tower (Download)[https://www.macintoshrepository.org/3915-simtower-cheats?ns=1] Download the .toast file and rename to .iso
- Ubuntu 12.04.5 (Download)[http://old-releases.ubuntu.com/releases/12.04.5/ubuntu-12.04.5-server-amd64.iso
- Marathon (Download)[https://macintoshgarden.org/sites/macintoshgarden.org/files/games/marathoncd.zip]

```
pi@rascsi:~/images $ sha1sum simtower.iso
ea6523a1b0c6d10181582b331912c21934bab317  simtower.iso
pi@rascsi:~/images $ sha1sum ubuntu-12.04.5-server-amd64.iso
7540ace2d6cdee264432f5ed987236d32edef798  ubuntu-12.04.5-server-amd64.iso
pi@rascsi:~/images $ sha1sum marathon.iso
3063489d009e9def138cd501f73239c0328a21a0  marathon.iso
pi@rascsi:~/images $ ls -al marathon.iso ubuntu-12.04.5-server-amd64.iso simtower.iso
-rw-r--r-- 1 www-data www-data  35895296 Jul 26  2020 marathon.iso
-rwxrwxr-x 1 pi       pi        16947200 Sep  9  2020 simtower.iso
-rwxrwxr-x 1 pi       pi       722468864 Sep  9  2020 ubuntu-12.04.5-server-amd64.iso
pi@rascsi:~/images $

```

Thank you for helping to clean up the crappy documentation. I never imagined someone would try to recreate my setup!

**Comment by akuker** [comment]
When you do get the robot tests running, note that the "Anex86 Hard Drive" always fails. I haven't taken the time to investigate, since I'm not aware of anyone using that functionality.

The first "Robot.Rascsi Cd-Rom Functionality" test also intermittently fails. It will usually pass the second time its run.

**Comment by uweseimet** [comment]
Thank you for providing setup details. Maybe the Wiki should be updated accordingly.

Talking about setups, do you see any issue with rasing the C++ level to c++-17 or, or even to c++-20? I'm using either current 32 bit versions of Ubuntu or Raspberry Pi OS, and it looks as if both levels work find with the compilers of these distributions. In case you agree to raise the level I will test the result on both environments, of course.

**Comment by akuker** [comment]
No objections to raising the c++ version! The official raspberry pi OS is based on a slightly older version of Debian. Not sure if it goes up to 20

**Comment by uweseimet** [comment]
OK! I am going to check which level the currrent official OS compiler supports.

**Comment by akuker** [comment]
When I ran the robot tests on these changes, a bunch of the cdrom tests fail. Does the CD-ROM functionality work on your end?

**Comment by uweseimet** [comment]
I can only run tests with hard disk and MO drives because currently I don't have a working CDROM setup on my retro hardware.
WIthout these changes the same tests run without any issues? If they do please provide the logfile of the failed tests, so that I can have a look at it.
You are sure that the tests are correct? Maybe the fail because of *correct* SCSI error codes being returned now?

**Comment by uweseimet** [comment]
I have set up my retro hardware for CD usage now and there are no issues with accessing data CDs.

**Comment by uweseimet** [comment]
@akuker Any news on this? I have not found any issues yet and tried 3 different operating systems. Have you already had the time to check whether the tests are correct?

By the way, I also tried to test with a Performa 630, but the hardware causes the Performa to hang when booting, see https://github.com/akuker/RASCSI/issues/132.

**Comment by akuker** [comment]
@uweseimet - Sorry, life has been busy this week. I haven't had a chance to test it, but if its working on a Performa 630 for you, I'll take that as good enough. I'll do my own testing before pushing it to the master branch.

**Review by akuker — APPROVED** [review]

**Comment by uweseimet** [comment]
Sounds good! Remember, the changes only affect the behavior in cases where an error (CHECK CONDITION) occured, i.e. the sense data returned by REQUEST SENSE after that. The behavior in the non-error case is unchanged.

**Pull request merged**